src/rust/src/utils/mod.rs

```diff
@@ -501,28 +528,79 @@ pub fn robj_to_usize(robj: extendr_api::Robj) -> RResult<usize> {
     robj_to_u64(robj).and_then(try_u64_into_usize)
 }
 
+fn err_no_nan<T>() -> RResult<T> {
+    rerr().plain("any NA value is not allowed here".to_string())
+}
+
+fn err_no_scalar<T>() -> RResult<T> {
+    rerr().plain("only a scalar value is allowed here (length = 1)".to_string())
+}
+
 pub fn robj_to_i64(robj: extendr_api::Robj) -> RResult<i64> {
     let robj = unpack_r_result_list(robj)?;
     use extendr_api::*;
+
+    return match (robj.rtype(), robj.len()) {
+        (_, 0 | 2..) => Some(err_no_scalar()),
+        (Rtype::Strings, 1) => Some(robj_parse_str_to_t(robj.clone())),
+        (Rtype::Doubles, 1) if robj.inherits("integer64") => {
+            robj_bit64_to_opt_i64(robj.clone()).map(Ok)
+        }
+        (Rtype::Doubles, 1) => robj.as_real().map(try_f64_into_i64),
+        (Rtype::Integers, 1) => robj.as_integer().map(i64::from).map(Ok),
+
+        (_, _) => {
+            Some(rerr().plain("does not support this R type for this conversion".to_string()))
+        }
+    }
+    .unwrap_or_else(err_no_nan)
+    .bad_robj(&robj)
+    .mistyped(tn::<i64>())
+    .when("converting into type");
+}
+
+pub fn robj_to_i32(robj: extendr_api::Robj) -> RResult<i32> {
+    let robj = unpack_r_result_list(robj)?;
+    use extendr_api::*;
+
     return match (robj.rtype(), robj.len()) {
-        (Rtype::Strings, 1) => robj
-            .as_str()
-            .unwrap_or("<Empty String>")
-            .parse::<i64>()
-            .ok(),
-        //specialized integer64 conversion
-        (Rtype::Doubles, 1) if robj.inherits("integer64") => robj
-            .as_real()
-            .and_then(|v| i64::try_from(v.to_bits()).ok())
-            .filter(|val| *val != crate::utils::BIT64_NA_ECODING),
-        //from R doubles or integers
-        (Rtype::Doubles, 1) => robj.as_real().and_then(|v| try_f64_into_i64(v).ok()),
-        (Rtype::Integers, 1) => robj.as_integer().map(i64::from),
-        (_, _) => None,
+        (_, 0 | 2..) => Some(err_no_scalar()),
+        (Rtype::Strings, 1) => Some(robj_parse_str_to_t(robj.clone())),
+        (Rtype::Doubles, 1) if robj.inherits("integer64") => {
+            robj_bit64_to_opt_i64(robj.clone()).map(try_i64_into_i32)
+        }
+        (Rtype::Doubles, 1) => robj.as_real().map(try_f64_into_i32),
+        (Rtype::Integers, 1) => robj.as_integer().map(i32::from).map(Ok),
+        (_, _) => {
+            Some(rerr().plain("does not support this R type for this conversion".to_string()))
+        }
     }
-    .ok_or(RPolarsErr::new())
+    .unwrap_or_else(err_no_nan)
     .bad_robj(&robj)
-    .mistyped(tn::<i64>());
+    .mistyped(tn::<i32>())
+    .when("converting into type");
+}
```

Comment:
Maybe consider about reusing the codes for `robj_to_i64` with additional bound checks?

Reply:
Maybe something like `Ok(i32::try_from(robj_to_i64(robj)?)?)`

Reply:
that is certainly also an option and I did so for usize. Errors can look weird though for the user if the error is about i64 but the function needed an i32. I tried to boil the conversion so it would not be to lengthy to do for each type, at least for the regulars f64, i64 i32.

Reply:
I must admit though, I do think I have gotten myself into to many rabbit holes lately and not progressed enough. Sry bout that :)